### PR TITLE
Refactor Notebook Link Handling

### LINKED ISSUE #15664: Fully qualified links inserted in WYSIWYG mode are turned into relative links
1. Create Notebook
2. Add text cell with link to local file
3. The link is turned into a relative path

e.g. 

`C:\Users\chgagnon\Desktop\test.txt` turns into `..\..\..\..\Desktop\test.txt`

This does not happen in split/MD mode.

### PR DESCRIPTION
**Pull request by VasuBhog**
This PR fixes #15664 and #15662.

The PR adds an additional setting `Keep Absolute Path` which when enabled ensures that if a user enters an absolute path it will stay when changing from WYSIWYG to Split/Markdown views. If the user enters a relative path with the setting enabled we ensure it stays relative as well.

The two cases `string` and `object` are triggered in different scenarios.

The first case `string` is triggered when a user enters a link to the linkCalloutDialog and we refactored it so the logic is now in one place. The returning output is the absolute path of the file or the web address link in string form. Once returned the string gets inserted as the href for the HTML Anchor Element [here](https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts#L254). This, in turn, will trigger the turndown service to call the getLinkURL (notebookLinkHandler) second case `object` with HTMLAnchorElement as such it will go through the conditions to check if it's an AnchorLink. Then last but not least will find the relative path from the current notebook directory to the target notebook user linked. The only time we would return absolute path is if the `Keep Absolute Path` setting is on.

### CODE REVIEW

File: src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts [hunk 1]
```diff
@@ -138,13 +139,12 @@ export class NotebookMarkdownRenderer {
...
 				if (uri.scheme !== 'mailto') {
 					href = href.replace(/&(?!amp;)/g, '&amp;');
-				} else {
-					href = href.replace(/</g, '&lt;')
-						.replace(/>/g, '&gt;')
-						.replace(/"/g, '&quot;')
-						.replace(/'/g, '&#39;');
 				}
-				return `<a href=${href} data-href="${href}" title="${title || href}">${text}</a>`;
+				href = href.replace(/</g, '&lt;')
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts:143]: Is this actually related to this PR? It looks like this was just a mistake from the last PR that got in

File: src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			let absolutePath = !path.isAbsolute(node.attributes.href?.nodeValue) ? path.resolve(path.dirname(notebookUri.fsPath), node.attributes.href?.nodeValue) : node.attributes.href?.nodeValue;
+			// if section link is different from the current notebook
+			return URI.file(absolutePath);
+		} else {
+			// else build an uri using the current notebookUri
+			return URI.from({ scheme: 'file', path: notebookUri.path, fragment: node.attributes.href?.nodeValue });
+		}
+	}
+
+	public get isAbsolutePath(): boolean {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts:107]: Could just make the property directly `public readonly isAbsolutePath`

File: src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+const keepAbsolutePathConfigName = 'notebook.keepAbsolutePath';
+
+export class NotebookLinkHandler {
+	private _notebookUriLink: URI | undefined;
+	private _href: string | undefined;
+	private _notebookDirectory: string;
+	private _isAnchorLink: boolean | undefined;
+	private _isFile: boolean | undefined;
+	private _isAbsolutePath: boolean | undefined;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts:14]: All these boolean values don't look like they can actually be undefined since you assign them in the constructor.

File: src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts (hunk 1 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts:143]: I saw that one of the commits didn't go through via the last PR. I saw this as I was refactoring so wanted to make sure the mistake was fixed.

Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts:143]: It's generally better to put fixes like that in separate PRs so they don't get lost among other changes. Makes it easier to review both sets of changes then.

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts:143]: Gotcha will revert this and put it in a separate PR.

File: src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
+import { URI } from 'vs/base/common/uri';
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts:1]: copyright header

(did it not warn you about missing this when committing? That should be done in the precommit hooks I thought)

File: src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts [hunk 5]
```diff
@@ -177,7 +177,7 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
 		this.updateTheme(this.themeService.getColorTheme());
 		this.setFocusAndScroll();
 		this.cellModel.isEditMode = false;
-		this._htmlMarkdownConverter = new HTMLMarkdownConverter(this.notebookUri);
+		this._htmlMarkdownConverter = new HTMLMarkdownConverter(this.notebookUri, this._configurationService);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts:180]: I guess this is fine since we already have the configuration service - but generally for proper DI usage we should be creating objects using instantiation service 

`this._htmlMarkdownConverter = this._instantiationService.createInstance(HTMLMarkdownConverter, this.notebookUri)`

**Review by Charles-Gagnon — COMMENTED** [review]
I didn't really go over

File: src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts (hunk 4 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts:1]: Never got a warning about it. Will add it though.

**Review by barbaravaldez — COMMENTED** [review]
Overall, it looks way better now! Thank you so much. I'll make sure to test as well 😄

File: src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	constructor(
+		private _notebookURI: URI,
+		private _link: string | HTMLAnchorElement,
+		@IConfigurationService private _configurationService: IConfigurationService,
+	) {
+		if (typeof this._link === 'string') {
+			this._notebookUriLink = URI.parse(this._link);
+			this._isFile = this._notebookUriLink.scheme === 'file';
+			this.isAbsolutePath = path.isAbsolute(this._link);
+			this._isAnchorLink = this._link.includes('#') && this._isFile;
```
[the remark this answers is not in the document]
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts:25]: Verified that when parsing an anchor link to a section we will get a file scheme. It works as expected.

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts [hunk 7]
```diff
@@ -132,27 +134,8 @@ export class HTMLMarkdownConverter {
...
-					if (relativePath) {
-						return `[${node.innerText}](${relativePath})`;
-					}
-				} else if (notebookLink?.fragment) {
-					// if the anchor link is to a section in the same notebook then just add the fragment
-					return `[${content}](${notebookLink.fragment})`;
-				}
-
+				const linkHandler = new NotebookLinkHandler(this.notebookUri, node, this.configurationService);
+				const href = linkHandler.getLinkUrl();
```
[the remark this answers is not in the document]
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:138]: Yup this works!

File: src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+	 * to the object case in which we will find the relative path of the file unless the user has the
+	 * keep absolute setting enabled then we don't convert absolute paths to relative paths
+	 * @returns the file link or web link
+	 */
+	public getLinkUrl(): string {
+		// cases where we only have the href link
+		if (typeof this._link === 'string') {
+			// Does not convert absolute path to relative path
+			if (this._isFile && this.isAbsolutePath && this._configurationService.getValue(keepAbsolutePathConfigName) === true) {
+				return this._link;
```
Comment (barbaravaldez) [inline src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts:62]: I noticed that by adding an absolute path with spaces and the keepAbsolute setting set to true the markdown link is not constructed properly. However, I do not see this issue when the keepAbsolute setting is set to false, because the relative link adds the %20 instead. 
`[/Users/{user}/hybrid book/content/readme.md](/Users/{user}/hybrid book/content/readme.md)`

**Review by barbaravaldez — APPROVED** [review]

File: src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts (hunk 8 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts:62]: Fixed by escaping the href when the setting is on but will fix the other %20 issues in the other PR #16422.

File: src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts [hunk 9]
```diff
@@ -108,6 +108,7 @@ export class NotebookMarkdownRenderer {
 			return '<img ' + attributes.join(' ') + '>';
 		};
 		renderer.link = (href: string, title: string, text: string): string => {
+			let hrefAbsolute: boolean = path.isAbsolute(href);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/outputs/notebookMarkdown.ts:111]: I'd add a comment here about why you're doing this at the top instead of down where it's used, assuming that the stuff we do between is the reason for why this is the case.

**Review by Charles-Gagnon — APPROVED** [review]
Didn't go over the logic in detail, rest of it looks good though.

**Review by corivera — APPROVED** [review]
